ingestion-beam/src/main/java/com/mozilla/telemetry/decoder/ParsePayload.java

```diff
@@ -158,44 +148,50 @@ protected PubsubMessage processElement(PubsubMessage message)
     return new PubsubMessage(normalizedPayload, attributes);
   }
 
-  private void addOperatingSystemAttributes(Map<String, String> attributes, JSONObject json) {
-    String osName = null;
-    String osVersion = null;
-
-    // Try to get "common ping"-style os fields.
-    JSONObject osObject = null;
-    try {
-      osObject = json.getJSONObject("environment").getJSONObject("system").getJSONObject("os");
-    } catch (JSONException ignore) {
-      // pass
-    }
-    if (osObject != null) {
-      Object name = osObject.opt("name");
-      if (name != null) {
-        osName = name.toString();
-      }
-      Object version = osObject.opt("version");
-      if (version != null) {
-        osVersion = version.toString();
-      }
+  private void addAttributesFromPayload(Map<String, String> attributes, JSONObject json) {
+
+    // Try to get glean-style client_info object.
+    Optional<JSONObject> gleanClientInfo = Optional.of(json) //
+        .map(j -> j.optJSONObject("client_info"));
+
+    // Try to get "common ping"-style os object.
+    Optional<JSONObject> commonPingOs = Optional.of(json) //
+        .map(j -> j.optJSONObject("environment")) //
+        .map(j -> j.getJSONObject("system")) //
```

Comment:
to prevent future headaches if a valid schema exists that doesn't match this shape I think we should make each nested key optional
```suggestion
        .map(j -> j.optJSONObject("system")) //
```

Reply:
That was my intention. Will change.